File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
from collections import deque

m,n,k = list(map(int, input().split()))

graph = [[0]*n for _ in range(m)]
visited = [[0]*n for _ in range(m)]

problem = [list(map(int, input().split())) for _ in range(k)]
for x1,y1,x2,y2 in problem:
  for y in range(x1, x2):
    for x in range(y1, y2):
      graph[x][y] = 1

dx = [0,0,-1,1]
dy = [-1,1,0,0]
solve =[]
def bfs(x,y):
  queue = deque()
  queue.append((x,y))
  visited[x][y] = 1
  count = 1
  while queue:
    x,y = queue.popleft()
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]
      
      if nx < 0 or ny < 0 or nx >= m or ny >= n:
        continue
      if graph[nx][ny] == 1 or visited[nx][ny] == 1:
        continue
      queue.append((nx, ny))
      visited[nx][ny] = 1
      count += 1
  return count
result = 0
for x in range(m):
  for y in range(n):
    if graph[x][y] == 0 and visited[x][y] == 0:
      solve.append(bfs(x,y))
      result += 1      
solve.sort()
print(result)
print(*solve)
  
  

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class Main {
  static int[][] graph;
  static boolean[][] visited;
  static int[] dx = new int[]{1,0,-1,0};
  static int[] dy = new int[]{0,1,0,-1};
  static int solCnt = 1;

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    String[] prob = br.readLine().split(" ");

    int xlen = Integer.parseInt(prob[1]);
    int ylen = Integer.parseInt(prob[0]);
    int cnt = Integer.parseInt(prob[2]);
    int solveCount = 0;
    ArrayList<Integer> sol = new ArrayList<>();


    graph = new int[xlen][ylen];
    visited = new boolean[xlen][ylen];

    for(int x=0; x <xlen; x++) {
      for(int y=0; y<ylen; y++) {
        graph[x][y] = 1;
      }
    }

    for(int i =0; i< cnt; i++) {
      String[] coord = br.readLine().split(" ");

      int[] fircoord = new int[]{Integer.parseInt(coord[0]), Integer.parseInt(coord[1])};
      int[] seccoord = new int[]{Integer.parseInt(coord[2]), Integer.parseInt(coord[3])};


      for(int x=fircoord[0]; x < seccoord[0]; x++) {
        for(int y=fircoord[1]; y< seccoord[1];y++) {
          graph[x][y] = 0;
        }
      }
    }

    for(int x = 0; x < xlen; x++) {
      for(int y =0; y<ylen; y++) {
        if (graph[x][y] == 1 && !visited[x][y]) {
          dfs(x,y,xlen, ylen);
          solveCount++;
          sol.add(solCnt);
          solCnt = 1;
        }
      }
    }

    System.out.println(solveCount);
    Collections.sort(sol);
    for(int q: sol) {
      System.out.print(q);
      System.out.print(" ");
    }

  }

  public static void dfs(int x, int y, int xlen, int ylen) {
    visited[x][y] = true;

    for(int i=0; i<4; i++) {
      int newx = x + dx[i];
      int newy = y + dy[i];

      if(newx>=0 && newy>=0 && newx< xlen && newy<ylen) {
        if(graph[newx][newy] == 1 && !visited[newx][newy]) {
          solCnt++;
          dfs(newx, newy, xlen, ylen);
        }
      }
    }
  }


}
